File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int ans = 0;
        int row = grid.size(), col = grid[0].size();
        vector<pair<int,int>> choices = {{0,-1},{0,1},{1,0},{-1,0}};
        for(int i =0; i<row; i++){
            for(int j = 0; j<col; j++){
                if(grid[i][j]=='0') continue;
                // here grid[i][j] = '1'
                ans++;
                queue<pair<int,int>> q; // row - col
                q.emplace(i,j); 
                grid[i][j] = '0';
                while(!q.empty()){
                    int curr_row = q.front().first;
                    int curr_col = q.front().second;
                    q.pop();
                    for(pair<int,int>& i : choices){
                        int r = curr_row + i.first, c = curr_col + i.second;
                        if(r<0 || c <0 || r>=row || c>=col || grid[r][c] == '0') continue;
                        grid[r][c] = '0';
                        q.emplace(r,c);
                    }
                }
            }
        }
        return ans;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    const offset = [ 0,1,0,-1,0];
    let ans = 0;
    const [row,col] = [grid.length, grid[0].length];
    const dfs = function (i,j) {
        grid[i][j] = '0';
        for(let k = 0; k<4; k++){
            const [new_r,new_c] = [i+offset[k],j+offset[k+1]];
            if(new_r<0 || new_c < 0 || new_r>=row || new_c >= col || grid[new_r][new_c] == '0') continue;
            dfs(new_r,new_c);
        }
    }
    for(let i = 0; i<row; i++){
        for(let j = 0; j<col; j++){
            if(grid[i][j]==='1'){
                ans++;
                dfs(i,j)
            }
        }
    }
    return ans;
};